Simplify maxcube integration

unaiur opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Device objects returned by maxcube-api dependency are stable, so
we do not need to resolve them from the device address every time.

Also, refactor and unify how maxcube integration sets temperature & mode.


## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
maxcube:
  gateways:
    - host: 192.168.178.26
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

unaiur committed:
Simplify maxcube integration

Device objects returned by maxcube-api dependency are stable, so
we do not need to resolve from the device address every time.

Also, refactor and unify how maxcube integration sets temperature & mode.

MartinHjelmare commented on homeassistant/components/maxcube/binary_sensor.py:
```diff
@@ -26,13 +23,12 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
-        self._name = name
-        self._sensor_type = DEVICE_CLASS_WINDOW
-        self._rf_address = rf_address
+        room = handler.cube.room_by_id(device.room_id)
```
Is this just a lookup of cached data or is it doing I/O?

unaiur replied:
handler.update() fetches all the state information. All getters returns the cached information.

MartinHjelmare commented on homeassistant/components/maxcube/binary_sensor.py:
```diff
@@ -42,15 +38,13 @@ def name(self):
     def is_on(self):
         """Return true if the binary sensor is on/open."""
-        return self._state
+        return self._device.is_open
```
Does this do I/O?

unaiur replied:
No, reads the cached state.

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -199,49 +172,30 @@ def hvac_action(self):
-            except (socket.timeout, OSError):
-                _LOGGER.error("Setting target temperature failed")
-                return False
+        self._set_target(None, kwargs.get(ATTR_TEMPERATURE))
```
The temperature parameter may be `None` so we should guard against that before calling this method. We usually raise a `ValueError` with appropriate message as exception argument on invalid service input.

unaiur replied:
In that case, it will call cube.set_temperature_mode(None, None), which does nothing. But I will change it to raise an exception.

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -257,37 +211,25 @@ def preset_modes(self):
+        elif preset_mode == PRESET_ON:
+            self._set_target(MAX_DEVICE_MODE_MANUAL, ON_TEMPERATURE)
+        elif preset_mode == PRESET_AWAY:
+            self._set_target(MAX_DEVICE_MODE_VACATION, None)
```
`None` is not a float so either the typing needs to change on `self._set_target` or we need to call it differently.

unaiur replied:
Both parameters of _set_target are optional. mode parameter defaults to current mode and temperature defaults to current temperature except when the mode is auto; in that case it read the temperature from the weekly schedule.

I will document the method properly.

unaiur committed:
Raise ValueError if missing parameters for set_temperature method

Raise a ValueError exception If set_temperature does not receive
a temperature parameter.

Also, document properly _set_target method.

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -151,37 +130,44 @@ def hvac_modes(self):
+        elif hvac_mode == HVAC_MODE_AUTO:
+            self._set_target(MAX_DEVICE_MODE_AUTOMATIC, None)
+
+    def _set_target(self, mode: Optional[int], temp: Optional[float]):
```
```suggestion
    def _set_target(self, mode: int | None, temp: float | None) -> None:
```

unaiur replied:
Done

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -1,6 +1,7 @@
 """Support for MAX! Thermostats via MAX! Cube."""
 import logging
```
```suggestion
from __future__ import annotations

import logging
```

unaiur committed:
Use Type | None instead of Optional[Type] annotation

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -151,37 +131,44 @@ def hvac_modes(self):
+            temp = max(self._device.target_temperature, self.min_temp)
+            self._set_target(MAX_DEVICE_MODE_MANUAL, temp)
+        elif hvac_mode == HVAC_MODE_AUTO:
+            self._set_target(MAX_DEVICE_MODE_AUTOMATIC, None)
```
Would be good to raise a ValueError if an hvac_mode is passed that isn't supported. Ie handle the `else:` case.

unaiur replied:
Done

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -257,37 +230,25 @@ def preset_modes(self):
+        elif preset_mode == PRESET_BOOST:
+            self._set_target(MAX_DEVICE_MODE_BOOST, None)
+        elif preset_mode == PRESET_NONE:
+            self._set_target(MAX_DEVICE_MODE_AUTOMATIC, None)
```
Here as well.

unaiur replied:
Done

unaiur committed:
Protect set_hvac_mode and set_preset_mode from unsupported parameters

MartinHjelmare reviewed (approved):
Looks good!

unaiur commented:
> Looks good!

My next CR will be moving to asyncio this integration... in a month or so I guess.

MartinHjelmare commented:
The device api in Home Assistant requires config entries. So I'd recommend:

1. AsyncIO
2. Config entries
3. Device info
